### PR TITLE
Add tagging to image sharer

### LINKED ISSUE #7: Image Tags
#### As a user I want to add tags to images.

__Story__:

Our users are loving the ability to add images to our service. However, now
that there are more than a handful of images, it is apparent that we need some
way to filter images by certain categories. In order to do that, we first need
to label images in some way. It is time to add the ability for our users to tag
images when they add them.

__Acceptance criteria__:
- [x] When I add a new image, there is a form field to add tags.
- [x] After I've saved an image with tags, I see those tags displayed with the
  associated image.
- [x] My application uses the acts-as-taggable-on gem to provide the tagging
  functionality.

__Reference material__:
- https://github.com/mbleigh/acts-as-taggable-on

__Dependencies__:
- Image Index

### PR DESCRIPTION
**Pull request by mathewbaltes**
Closes #7

### CODE REVIEW

File: db/migrate/20170404215107_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# This migration comes from acts_as_taggable_on_engine (originally 1)
+class ActsAsTaggableOnMigration < ActiveRecord::Migration
+  def self.up
+    create_table :tags do |t|
+      t.string :name
```
Comment (shfung) [inline db/migrate/20170404215107_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb]: Maybe we can add some not-null constraints to some of the columns?

File: app/views/images/new.html.erb [hunk 2]
```diff
@@ -1,7 +1,8 @@
 <%= simple_form_for @image, url: images_path do |f| %>
   <div class="row justify-content-center">
     <div class="col-lg-6">
-      <%= f.input :url, label: false, as: :string, input_html: { id: 'link-input', placeholder: 'Enter a image link' } %>
+      <%= f.input :url, label: 'Image link', as: :string, input_html: { id: 'link-input', placeholder: 'Enter a image link' } %>
+      <%= f.input :tag_list, label: 'Tags [Optional]', input_html: { id: 'tag-input', placeholder: 'Enter comma separated tags' } %>
```
Comment (shfung) [inline app/views/images/new.html.erb]: If you submit the form with an invalid URL and comma separated tags, are the tags still comma separated after you got back an error?

File: app/models/image.rb [hunk 3]
```diff
@@ -1,4 +1,6 @@
 class Image < ApplicationRecord
   include ActiveModel::Validations
```
Comment (shfung) [inline app/models/image.rb]: Not related to this issue: I don't think you need to `include ActiveModel::Validations` if this class is a ActiveRecord

File: app/views/images/show.html.erb [hunk 4]
```diff
@@ -6,8 +6,8 @@
   </div>
 </div>
 <div class="row py-1 justify-content-center">
-  <div class="col text-center">
-    <%= image_tag @image.url, id: 'submitted-image', class: 'shared-image img-responsive'%>
+  <div class="col text-center shared-image-block">
+    <%= image_tag @image.url, id: 'submitted-image', class: 'img-responsive'%>
```
Comment (shfung) [inline app/views/images/show.html.erb]: Do we want to show the tags in the show page?

File: test/controllers/images_controller_test.rb [hunk 5]
```diff
@@ -24,6 +24,7 @@ def setup
       assert_select "img[src='#{existing_image.url}']"
     end
     assert_select 'img', count: 2
+    assert_select '.badge', count: 1
```
Comment (shfung) [inline test/controllers/images_controller_test.rb]: Maybe also assert that the correct tag is shown?

File: app/models/image.rb (hunk 3 above)
Comment (mathewbaltes) [inline app/models/image.rb]: Very nice catch!

File: app/views/images/new.html.erb (hunk 2 above)
Comment (mathewbaltes) [inline app/views/images/new.html.erb]: Nope, but they will be now. Thanks!

File: db/migrate/20170404215107_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb [hunk 6]
```diff
@@ -0,0 +1,31 @@
+# This migration comes from acts_as_taggable_on_engine (originally 1)
+class ActsAsTaggableOnMigration < ActiveRecord::Migration
+  def self.up
+    create_table :tags do |t|
+      t.string :name, null: false
+    end
+
+    create_table :taggings do |t|
+      t.references :tag
```
Comment (shfung) [inline db/migrate/20170404215107_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb]: There are more columns in these two tables that should have not-null constraints.

File: test/models/images_model_test.rb [hunk 7]
```diff
@@ -2,18 +2,32 @@
 
 class ImageModelTest < ActiveSupport::TestCase
   test 'verify image with valid url is valid' do
-    image = Image.create(url: 'http://www.appfolio.com/')
+    image = Image.new(url: 'http://www.appfolio.com/', tag_list: 'blah')
     assert_equal true, image.valid?
     assert_empty image.errors
```
Comment (shfung) [inline test/models/images_model_test.rb]: Maybe add a assertion that the tag list is being set correctly?

File: test/controllers/images_controller_test.rb [hunk 8]
```diff
@@ -38,30 +39,55 @@ def setup
...
   test 'once link is created we can see details about the new link' do
-    post images_path, params: { image: { url: @test_src } }
+    test_tag_list = 'blah, dee, boop'
+    assert_difference 'Image.count', 1 do
+      post images_path, params: { image: { url: @test_src, tag_list: test_tag_list } }
+    end
+
+    created_image = Image.last
 
     follow_redirect!
```
Comment (shfung) [inline test/controllers/images_controller_test.rb]: I would suggest not calling `follow_redirect!` here. There are two reasons:
1. each test in the controller test should only fire one request.
2. testing content of the show page in a test mainly for the new action can be confusing.

Instead, you should call `assert_redirected_to` to assert on the response of the controller action.

File: app/views/images/show.html.erb (hunk 4 above)
Comment (mathewbaltes) [inline app/views/images/show.html.erb]: Sure do, I've added the tags.

**Pull request merged**